### PR TITLE
add some more interfaces to BaseInterfaces.jl

### PR DESCRIPTION
**Pull request by rafaqz**
This PR adds the start of `AbstractSet`, `AbstractDict`, and `AbstractArray` to BaseInterfaces.jl

Hopefully this will end with interface tests that help people write interfaces for these base types and
know they are correct and complete.

@gdalle thoughts or contributions to the cause would be appreciated!

### CODE REVIEW

**Comment by gdalle** [comment]
Wow, large effort! I'll try to review it in the coming days. Just wondering, did you check out other initiatives like https://github.com/JuliaArrays/ArrayInterface.jl?

**Comment by rafaqz** [comment]
Yeah, I've been following ArrayInterface for some years.

But ArrayInterface.jl really isn't the same as the base `AbstractArray` interface. As the docs say: 

> The purpose of this library is to solidify *extensions* to the current AbstractArray interface

(emphasis mine)

The idea in BaseInterfaces.jl is just to have a simple definitions of Base Julia interfaces as far as they are documented, or as used in Base where they are not documented. Its a test of Interfaces.jl capacity to represent a range of interfaces, and should be useful for people extending theses interfaces to have a predefined test suit for it. I'm not sure that the `@implements` declarations will actually be used or useful, but that's ok.

Of course, there will be some overlap in terms of traits as ArrayInterface.jl does include some traits for simple base interface things as well.

**Comment by gdalle** [comment]
Feel free to ping me when this is more or less ready for review!

**Comment by rafaqz** [comment]
Any time really, feedback would be good. It will be a work in progress untul we register, so we can merge PRs as we go.

**Comment by gdalle** [comment]
Okay cool, I'll try to take a look

File: BaseInterfaces/src/array.jl [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        ),
+        size = (
+            A -> size(A) isa NTuple{<:Any,Int},
+            A -> length(size(A)) == ndims(A),
+        ),
+        getindex = (
+            A -> A[begin] isa eltype(A),
+            A -> A[map(first, axes(A))...] isa eltype(A),
+        ),
+        indexstyle = A -> IndexStyle(A) in (IndexCartesian(), IndexLinear()),
```
Comment (Seelengrab) [inline BaseInterfaces/src/array.jl:56]: This isn't mandatory - there is a default for this.

File: BaseInterfaces/src/array.jl [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    ),
+    # TODO implement all the optional conditions
+    optional = (;
+        setindex! = (
+            A -> length(A) > 1 || throw(ArgumentError("Test arrays must have more than one element to test setindex!")),
```
Comment (Seelengrab) [inline BaseInterfaces/src/array.jl:61]: I think a better property here is to check that setting any index from `eachindex` and retrieving that index again gives the same element. This isn't actually defined by Base though... In some contexts, this can be only temporarily true, e.g. if you have a `DistributedArray` that fails to propagate some information in an overlapping region due to a network failure, for example.

File: BaseInterfaces/src/dict.jl [hunk 3]
```diff
@@ -0,0 +1,17 @@
+
+@interface DictInterface AbstractDict (
+    mandatory = (;
```
Comment (Seelengrab) [inline BaseInterfaces/src/dict.jl:3]: What about `keys` and `values`?

File: BaseInterfaces/src/set.jl [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            typeof(sd) == typeof(s) && length(sd) == 0
+        end,
+        intersect = s -> intersect(s, s) == s, # TODO
+        union = s -> union(s, s) == s, # TODO
+        empty! =  "empty! removes all elements from the set" => s -> length(empty!(s)) == 0,
+        delete! = "delete! removes element valued x of the set" => s -> begin
+            x = first(iterate(s))
+            !in(delete!(s, x), x)
+        end,
+        push! = "push! adds element x to the set" => s -> begin
```
Comment (Seelengrab) [inline BaseInterfaces/src/set.jl:44]: This assumes non-emptiness.

File: BaseInterfaces/src/set.jl [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            x = first(iterate(s))
+            delete!(s, x)
+            push!(s, x)
+            in(x, s)
+        end,
+        copymutable = s -> begin
+            s1 = Base.copymutable(s)
+            s1 isa typeof(s) && s1 !== s
+        end,
+        # TODO is there anything we can actually test here?
```
Comment (Seelengrab) [inline BaseInterfaces/src/set.jl:55]: Unfortunately not, since there is no `capacity` function in Base :/

File: BaseInterfaces/src/set.jl [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            s1 = copy(s) 
+            s1 !== s && s1 isa typeof(s) && collect(s) == collect(s1)
+        end,
+        empty = "returns a empty set able to hold elements of type U" => s -> begin
+            s1 = Base.empty(s)
+            eltype(s1) == eltype(s) || return false
+            s1 = Base.empty(s, Int)
+            eltype(s1) == Int || return false
+        end,
+        emptymutable = "returns a empty set able to hold elements of type U" => s -> begin
```
Comment (Seelengrab) [inline BaseInterfaces/src/set.jl:25]: Also `isempty`

File: BaseInterfaces/src/set.jl [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        end,
+        hasfastin = s -> Base.hasfastin(s) isa Bool,
+        setdiff = s -> begin
+            sd = setdiff(s, collect(s))
+            typeof(sd) == typeof(s) && length(sd) == 0
+        end,
+        intersect = s -> intersect(s, s) == s, # TODO
+        union = s -> union(s, s) == s, # TODO
+        empty! =  "empty! removes all elements from the set" => s -> length(empty!(s)) == 0,
+        delete! = "delete! removes element valued x of the set" => s -> begin
```
Comment (Seelengrab) [inline BaseInterfaces/src/set.jl:40]: This assumes non-emptiness.

File: BaseInterfaces/src/array.jl [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+_ndims(::AbstractArray{<:Any,N}) where N = N
+
+array_components = (;
+    mandatory = (;
+        type = A -> A isa AbstractArray,
+        eltype = (
+            A -> eltype(A) isa Type,
+            A -> eltype(A) == _eltype(A),
+        ),
+        ndims = (
```
Comment (Seelengrab) [inline BaseInterfaces/src/array.jl:44]: This should have a fallback definition already, no? IIRC `ndims` comes from having `size` defined already.

**Review by Seelengrab — COMMENTED** [review]
Just some quick things I noticed :)

**Review by gdalle — COMMENTED** [review]
I see that @Seelengrab has reviewed the array, dict and set interfaces, so I'll submit my review on everything except those three files for now

File: BaseInterfaces/src/iteration.jl [hunk 9]
```diff
@@ -74,18 +70,19 @@ end
...
 
-@interface IterationInterface (
+@interface IterationInterface Any (
     # Mandatory conditions: these must be met by all types
     # that implement the interface.
     mandatory = (
         iterate = test_iterate,
-        size = test_size,
-        eltype = test_eltype,
+        isiterable = x -> Base.isiterable(typeof(x)),
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/src/iteration.jl:78]: Yes its not, but it should be. Its in there to make people implement it in future.

The docs are unfortunately not a reliable spurce of truth.

File: BaseInterfaces/test/runtests.jl [hunk 10]
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
 using Interfaces
 using Test
 
-@testset "BaseInterfaces.jl" begin
+@implements SetInterface{(:empty,:emptymutable,:hasfastin,:intersect,:union,:sizehint)} Test.GenericSet
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/test/runtests.jl:5]: I using `subtypes(AbstractSet)` to get everything I can find.

File: BaseInterfaces/src/BaseInterfaces.jl [hunk 11]
```diff
@@ -2,15 +2,43 @@ module BaseInterfaces
...
+export ArrayInterface, DictInterface, IterationInterface, SetInterface
 
 include("iteration.jl")
+include("dict.jl")
+include("set.jl")
+include("array.jl")
 
 # Some example interface delarations.
+
+# @implements ArrayInterface Base.LogicalIndex # No getindex
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/src/BaseInterfaces.jl:14]: Base just breaks the interface 😱

I have to special case it in DimensionalData.jl for this reason...

File: BaseInterfaces/test/runtests.jl [hunk 12]
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
...
+@implements DictInterface Test.GenericDict
+
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/test/runtests.jl:7]: I dont either :)

File: BaseInterfaces/test/runtests.jl [hunk 13]
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
...
+    # No `getindex` defined for LogicalIndex
+    @test_broken Interfaces.test(ArrayInterface, Base.LogicalIndex, [to_indices([1, 2, 3], ([false, true, true],))[1]])
+
+    # TODO test LinearAlgebra arrays and SparseArrays
+end
+
+@testset "DictInterface" begin
+    @test Interfaces.test(DictInterface, Dict, [Arguments(d=Dict(:a => 1, :b => 2), k=:c, v=3)])
+    @test Interfaces.test(DictInterface, IdDict, [Arguments(d=IdDict(:a => 1, :b => 2), k=:c, v=3)])
+    @test Interfaces.test(DictInterface, Base.EnvDict, [Arguments(d=Base.EnvDict())])
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/test/runtests.jl:29]: EnvDict fills itself with Env key/values

File: BaseInterfaces/src/iteration.jl [hunk 14]
```diff
@@ -74,18 +70,19 @@ end
 # `Iterators.reverse` gives reverse iteration 
 test_reverse(x) = collect(Iterators.reverse(x)) == reverse(collect(x))
 
-@interface IterationInterface (
+@interface IterationInterface Any (
     # Mandatory conditions: these must be met by all types
     # that implement the interface.
     mandatory = (
         iterate = test_iterate,
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/src/iteration.jl:77]: Yes often we need to force the test data to have a minimum length. 

But in this case we cant always call length on an iterator so not sure how to test it

File: BaseInterfaces/src/array.jl (hunk 1 above)
Comment (rafaqz) [reply BaseInterfaces/src/array.jl:56]: So it will still always work? Its saying its mandatory tgat it returns one of those options.

File: BaseInterfaces/src/dict.jl (hunk 3 above)
Comment (rafaqz) [reply BaseInterfaces/src/dict.jl:3]: Yeah Dict is not really fleshed out yet

File: BaseInterfaces/src/array.jl (hunk 8 above)
Comment (rafaqz) [reply BaseInterfaces/src/array.jl:44]: Doesnt it use the second type parameter of AbstractArray? 

I just wanted to check that was not messed up in the implementation of a custom type. Maybe not titally useful

File: BaseInterfaces/src/set.jl (hunk 6 above)
Comment (rafaqz) [reply BaseInterfaces/src/set.jl:25]: Somehow I forgot aboit `isempty` 😅

File: BaseInterfaces/test/runtests.jl [hunk 15]
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
...
+    @test Interfaces.test(ArrayInterface, SubArray, [view([7, 2], 1:2)])
+    @test Interfaces.test(ArrayInterface, PermutedDimsArray, [PermutedDimsArray([7 2], (2, 1))])
+    @test Interfaces.test(ArrayInterface, Base.ReshapedArray, [reshape(view([7, 2], 1:2), 2, 1)])
+    @test Interfaces.test(ArrayInterface, UnitRange, [2:10])
+    @test Interfaces.test(ArrayInterface, StepRange, [2:1:10])
+    @test Interfaces.test(ArrayInterface, Base.OneTo, [Base.OneTo(10)])
+    @test Interfaces.test(ArrayInterface, Base.Slice, [Base.Slice(100:150)])
+    @test Interfaces.test(ArrayInterface, Base.IdentityUnitRange, [Base.IdentityUnitRange(100:150)])
+    @test Interfaces.test(ArrayInterface, Base.CodeUnits, [codeunits("abcde")])
+    # No `getindex` defined for LogicalIndex
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/test/runtests.jl:20]: It only fails with e.g.  `CUDA.allowscalar(true)` on GPU. So thst would need to be `false`.

The docs really specificially say its one of the main mandatory methods, I kinda want the failed test there to demonstrate the problem...

File: BaseInterfaces/src/set.jl (hunk 4 above)
Comment (rafaqz) [reply BaseInterfaces/src/set.jl:44]: I added a !isempty test in mandatory

File: BaseInterfaces/src/set.jl (hunk 5 above)
Comment (rafaqz) [reply BaseInterfaces/src/set.jl:55]: Its so weird that the hint just disappears and we cant access it.

File: BaseInterfaces/src/set.jl (hunk 7 above)
Comment (rafaqz) [reply BaseInterfaces/src/set.jl:40]: Fixed as above by requiring the test data is not empty

File: BaseInterfaces/src/iteration.jl [hunk 16]
```diff
@@ -74,18 +70,19 @@ end
...
-        eltype = test_eltype,
+        isiterable = x -> Base.isiterable(typeof(x)),
+        size = test_iterator_size,
+        eltype = test_iterator_eltype,
     ),
     # Optional conditions. These should be specified in the
     # interface type if an object implements them: IterationInterface{(:reverse,:indexing)}
     optional = (
         reverse = test_reverse,
         indexing = test_indexing,
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/src/iteration.jl]: I think iterators like that just cant just define `getindex`.

File: BaseInterfaces/src/iteration.jl [hunk 17]
```diff
@@ -49,9 +49,10 @@ function test_size(x)
     end
 end
 
-function test_eltype(x)
+function test_iterator_eltype(x)
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/src/iteration.jl:45]: I refactored all of this to be inline instead, I think having the function outside its harder to understand the interface

File: BaseInterfaces/test/runtests.jl [hunk 18]
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
...
     @test Interfaces.test(IterationInterface, Base.Generator, [(i for i in 1:5), (i for i in 1:5)])
     @test Interfaces.test(IterationInterface, Tuple, [(1, 2, 3, 4)])
 end
+
+@testset "SetInterface" begin
+    @test Interfaces.test(SetInterface, Set, [Set((1, 2))])
+    @test Interfaces.test(SetInterface, BitSet, [BitSet((1, 2))])
+    @test Interfaces.test(SetInterface, Base.KeySet, [Base.KeySet(Dict(:a=>1, :b=>2))])
+    @test Interfaces.test(SetInterface, Test.GenericSet, [Test.GenericSet(Set((1, 2)))])
+    # @test Interfaces.test(SetInterface, Base.IdSet, ?)
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/test/runtests.jl]: Just annoying to construct. Fixed

File: BaseInterfaces/test/runtests.jl [hunk 19]
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
...
+
+    # TODO test LinearAlgebra arrays and SparseArrays
+end
+
+@testset "DictInterface" begin
+    @test Interfaces.test(DictInterface, Dict, [Arguments(d=Dict(:a => 1, :b => 2), k=:c, v=3)])
+    @test Interfaces.test(DictInterface, IdDict, [Arguments(d=IdDict(:a => 1, :b => 2), k=:c, v=3)])
+    @test Interfaces.test(DictInterface, Base.EnvDict, [Arguments(d=Base.EnvDict())])
+    @test Interfaces.test(DictInterface, Base.ImmutableDict, [Arguments(d=Base.ImmutableDict(:a => 1, :b => 2))])
+    # @test Interfaces.test(DictInterface, Base.Pairs, [Arguments(d=Base.pairs((a = 1, b = 2)))])
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply BaseInterfaces/test/runtests.jl]: Fixed

**Comment by rafaqz** [comment]
Thanks for the feedback, I've implemented most of these and a bunch more. Also fixed up testing and display a lot and accidentally merged it here! But will merge as-is anyway and more can be added in another PR.

File: BaseInterfaces/src/array.jl (hunk 8 above)
Comment (Seelengrab) [reply BaseInterfaces/src/array.jl:45]: True, but that's not an issue because you have to declare where in a subtype the second parameter of a parent type goes. E.g. it doesn't matter that I can swap the parameters around in my definition:

```
julia> struct MyArr{N,T} <: AbstractArray{T,N} end

julia> ndims(MyArr{2,Int})
2
```

File: BaseInterfaces/src/set.jl (hunk 5 above)
Comment (Seelengrab) [reply BaseInterfaces/src/set.jl]: Well it's just a hint/suggestion, not a requirement :shrug:
